src/audio_input/audio.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2020 Patryk Seregiet
+ *
+ * This file is part of ReplaySorcery.
+ *
+ * ReplaySorcery is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ReplaySorcery is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with ReplaySorcery.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "audio.h"
+#include "../util/log.h"
+#include "../util/memory.h"
+#define SAMPLES_PER_CALLBACK 4096 // has to be power of 2
+
+static void audioCallback(void *userdata, uint8_t *stream, int len) {
+   RSAudio *audio = (RSAudio *)userdata;
+   rsCircleStaticAdd(&audio->data, stream, len);
+}
+
+static void getDeviceName(RSAudio *audio, const RSConfig *config) {
+   const char *monitor = "Monitor of ";
+   const char *devname = config->audioDeviceName;
+   if (!strcmp(config->audioDeviceName, "unknown")) {
+      audio->deviceName = NULL;
+      return;
+   }
+   if (!strcmp(devname, "auto")) {
+      const char *sdlname = SDL_GetAudioDeviceName(0, false);
+      if (!sdlname) {
+         audio->deviceName = NULL;
+         return;
+      }
+      audio->deviceName = rsMemoryCreate(strlen(sdlname) + strlen(monitor) + 1);
+      strcpy(audio->deviceName, monitor);
+      strcat(audio->deviceName, sdlname);
+      return;
+   }
+   audio->deviceName = rsMemoryCreate(strlen(devname) + 1);
+   strcpy(audio->deviceName, devname);
```

Comment:
I should have just used strdup, does the same thing as rsStringClone
https://man7.org/linux/man-pages/man3/strdup.3.html

Reply:
I'm not sure, I've been trying to keep to mostly C99 std functions

Reply:
then why do you use dynamic size arrays ? like char somethng[xxx] where xxx in not constant. It's a GNU extension and will only work on gcc

Reply:
It isn't, it's actually one of the most useful and most unheard of features of C. See [here](https://en.cppreference.com/w/c/language/array) and scroll down to where it talks about "Variable Length Arrays" (VLAs). If it was a GCC extension I'm pretty sure `-Wpedantic` would've told me about it, since [that is what that warning flag is designed for](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html).

Reply:
Alright I was wrong then.

Reply:
done